collect_app/src/androidTest/java/org/odk/collect/android/preferences/qr/QrCodeActivitiesTest.java

```diff
@@ -0,0 +1,126 @@
+import dagger.Provides;
+import io.reactivex.Observable;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.junit.runner.RunWith;
+import org.odk.collect.android.R;
+import org.odk.collect.android.activities.MainMenuActivity;
+import org.odk.collect.android.injection.config.AppDependencyModule;
+import org.odk.collect.android.support.ResetStateRule;
+import org.odk.collect.android.support.pages.MainMenuPage;
+
+import java.util.Collection;
+
+import static androidx.test.espresso.intent.Intents.intended;
+import static androidx.test.espresso.intent.Intents.intending;
+import static androidx.test.espresso.intent.matcher.IntentMatchers.hasAction;
+import static androidx.test.espresso.intent.matcher.IntentMatchers.hasType;
+import static androidx.test.espresso.intent.matcher.IntentMatchers.isInternal;
+import static org.hamcrest.Matchers.not;
+
+
+@RunWith(AndroidJUnit4.class)
+public class QrCodeActivitiesTest {
+    // drawable resource that will act as "qr code" in this test
+    private final int checkerBackgroundDrawableId = R.drawable.checker_background;
+
+    @Rule
+    public IntentsTestRule<MainMenuActivity> rule = new IntentsTestRule<>(MainMenuActivity.class);
+
+    @Rule
+    public RuleChain copyFormChain = RuleChain
+            .outerRule(GrantPermissionRule.grant(
+                    Manifest.permission.READ_EXTERNAL_STORAGE,
+                    Manifest.permission.WRITE_EXTERNAL_STORAGE,
+                    Manifest.permission.READ_PHONE_STATE
+            ))
+            .around(new ResetStateRule(new AppDependencyModule() {
+                @Override
+                @Provides
+                public QRCodeGenerator providesQRCodeGenerator() {
+                    return new QRCodeGenerator() {
+                        @Override
+                        public Observable<Bitmap> generateQRCode(Collection<String> selectedPasswordKeys) {
+                            return Observable.create(emitter -> {
+                                Bitmap bitmap =
+                                        BitmapFactory.decodeResource(
+                                                ApplicationProvider.getApplicationContext().getResources(),
+                                                checkerBackgroundDrawableId);
+                                emitter.onNext(bitmap);
+                                emitter.onComplete();
+                            });
+                        }
+                    };
+                }
+            }));
+
+    @Before
+    public void stubAllExternalIntents() {
+        // By default Espresso Intents does not stub any Intents. Stubbing needs to be setup before
+        // every test run. In this case all external Intents will be blocked.
+        intending(not(isInternal())).respondWith(new Instrumentation.ActivityResult(Activity.RESULT_OK, null));
+    }
+
+    @Test
+    public void checkQRScannerIsInitiated() {
```

Comment:
Again just in terms of naming lets try and get closer to describing the feature. `onMainMenu_clickConfigureQRCode_andClickingOnScan_opensScanner()` for this as an example. I'd try doing a similar "given-when-then" style for all these tests.